### PR TITLE
a blog post on GC latencies in Haskell, OCaml, Racket

### PR DESCRIPTION
**Pull request by gasche**
Two things:
- I worry that the blog post could be too long and that I should maybe split it in two. Could people give me their feeling on this? (In my experience splitting posts is a bit of a pain, so it's only justified if we are confident people will skip the later parts of the post otherwise.)
- I could not check with the CSS enabled locally (the local index.html does not work very well for me) so I don't know whether the pygments coloration actually looks nice -- I checked that the produced HTML does use pygments classes. Is there a way to know that before publishing the post? In particular, on my machine I had to use `scheme` language name as `racket` was not supported (it's only in fairly recent versions, maybe?), can someone tell me whether `racket` coloring is supported on the machine of whoever makes the push decision?

### CODE REVIEW

**Comment by takikawa** [comment]
The `racket` coloring does work on my machine anyway, though I'm not really sure it looks nicer.

File: blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+
+(define (message n) (make-bytes 1024 (modulo n 256)))
+
+(define (push-msg chan id-high)
+  (define id-low (id-high . - . window-size))
+  (define inserted (hash-set chan id-high (message id-high)))
+  (if (id-low . < . 0) inserted
+      (hash-remove inserted id-low)))
+
+(define _
```
Comment (takikawa) [inline blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md:312]: Is this definition necessary? Is it just here to make it look similar to the OCaml code?

**Comment by takikawa** [comment]
I think the length is fine. I suspect some readers will skip portions, but that's ok. (e.g., some people may only be interested in the OCaml part or the Racket part) Nice post!

File: blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md (hunk 1 above)
Comment (gasche) [inline blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md:312]: Good point, I'll remove the `define` wrapping.

Comment (gasche) [inline blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md:312]: Hm I just tried and if I remove the `define`, racket will by default decide to print the resulting value, which is a immutable hashmap with 200,000 strings of lengths 1,024 -- not so nice!

Comment (takikawa) [inline blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md:312]: Oh I see. Keeping the `define` sounds fine then. Another thing you can do is pass the whole expression to `void`, like `(void (for/fold ...))` but that's not any nicer really.

File: blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+        racket -l gcstats -t main.rkt
+
+- By setting the environment variable `PLTSTDERR=debug@GC`, the racket
+  runtime will log GC events on the standard error output. One can
+  then grep for minor or major collections, or produce a histogram of
+  running times through the following scripting soup I cooked myself:
+
+        cat racket.log | grep -v total | cut -d' ' -f7 | sort -n | uniq --count
+
+Racket has an instrumental GC that is currently experimental (it is
```
Comment (bennn) [inline blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md]: "incremental GC"

Comment (bennn) [inline blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md]: "incremental GC"

File: blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+This is good news for our benchmark: sure, there is a very bad pause
+at the beginning of the program, but it's a one-time thing. It does
+not really affect the last decile of latency that is discussed in
+James Fischer's post, and would not be a problem during the steady
+state of an actual message-passing application.
+
+### Tuning the Racket version
+
+Matthew Flatt also remarked that by inserting explicit calls to the
+GC, one can get collection preformed more often than Racket's
```
Comment (bennn) [inline blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md]: performed

File: blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+heuristics demand and partly avoid the large peak pause. However,
+during explicit collection too frequently hurts the program
```
Comment (bennn) [inline blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md]: remove "during"?

**Comment by bennn** [comment]
very nice! Don't split it

**Comment by gasche** [comment]
Thanks @benn, I fixed the typos.

**Comment by gasche** [comment]
I committed the generated files¹. I think the post should be ready to publish whenever possible.

¹: This step is a bit cumbersome. Do you think it would be possible to transition to a system where we only commit sources to the main branch, and the generated content is automatically committed to a separate branch (by a `make blog` rule for example) at publication time? This way, authors would only deal with sources, which decreases the entry barrier, and editors would manually run builds to release (which also insures consistency, for example in term of Frog version used, pygments support, etc.), yet it would still all be in the git repository.

**Comment by bennn** [comment]
published in 35b9b7e9
(sorry about the delay)

**Comment by gasche** [comment]
I see no syntax highlighting on the final rendered blog post, http://prl.ccs.neu.edu/blog/2016/05/24/measuring-gc-latencies-in-haskell-ocaml-racket/ . @benn, is it necessary to explicitly include some pygments-provided css files somewhere in the blog templates?

**Comment by bennn** [comment]
I'll check (it's `bennn` with 3 n's BTW)

**Comment by bennn** [comment]
Do you have the `pygments.css` file you were using? (The one on your `blog-post-latencies` branch is the same as on master here ... did you use a different one?)

**Comment by bennn** [comment]
ok the real problem seems to be I don't know how to use pygments

**Comment by bennn** [comment]
The final post has colors now.

(Issue was, I had nothing symlinked to `python` in favor of links like `python2.7` and `python3.5`. Fixing the link and rebuilding the blog rendered the code blocks correctly)

**Pull request closed without merging**